internal/controllers/synthesis/lifecycle_test.go

```diff
@@ -674,14 +674,75 @@ func TestShouldSwapStates(t *testing.T) {
 
 func TestInputRevisionsEqual(t *testing.T) {
 	synth := &apiv1.Synthesizer{}
-	synth.Spec.Refs = []apiv1.Ref{{Key: "foo"}, {Key: "bar", Defer: true}}
+	synth.Spec.Refs = []apiv1.Ref{{Key: "foo"}, {Key: "bar", Defer: true}, {Key: "baz"}}
+
+	tcs := []struct {
+		name  string
+		a, b  []apiv1.InputRevisions
+		equal bool
+	}{
+		{
+			name:  "just keys",
+			a:     []apiv1.InputRevisions{{Key: "foo"}, {Key: "baz"}},
+			b:     []apiv1.InputRevisions{{Key: "foo"}, {Key: "baz"}},
+			equal: true,
+		},
+		{
+			name:  "resource version mismatch",
+			a:     []apiv1.InputRevisions{{Key: "foo"}, {Key: "baz"}},
+			b:     []apiv1.InputRevisions{{Key: "foo", ResourceVersion: "not-zero"}, {Key: "baz"}},
+			equal: false,
+		},
+		{
+			name:  "revision missong",
+			a:     []apiv1.InputRevisions{{Key: "foo"}, {Key: "baz"}},
+			b:     []apiv1.InputRevisions{{Key: "foo", Revision: ptr.To(123)}, {Key: "baz"}},
+			equal: false,
+		},
+		{
+			name:  "revision mismatch",
+			a:     []apiv1.InputRevisions{{Key: "foo", Revision: ptr.To(234)}, {Key: "baz"}},
+			b:     []apiv1.InputRevisions{{Key: "foo", Revision: ptr.To(123)}, {Key: "baz"}},
+			equal: false,
+		},
+		{
+			name:  "revision match",
+			a:     []apiv1.InputRevisions{{Key: "foo", Revision: ptr.To(123)}, {Key: "baz"}},
+			b:     []apiv1.InputRevisions{{Key: "foo", Revision: ptr.To(123)}, {Key: "baz"}},
+			equal: true,
+		},
+		{
+			name:  "resource version match",
+			a:     []apiv1.InputRevisions{{Key: "foo", ResourceVersion: "not-zero"}, {Key: "baz"}},
+			b:     []apiv1.InputRevisions{{Key: "foo", ResourceVersion: "not-zero"}, {Key: "baz"}},
+			equal: true,
+		},
+		{
+			name:  "mixed resource version and revision",
+			a:     []apiv1.InputRevisions{{Key: "foo", ResourceVersion: "not-zero"}, {Key: "baz"}},
+			b:     []apiv1.InputRevisions{{Key: "foo", Revision: ptr.To(123)}, {Key: "baz"}},
+			equal: false,
+		},
+		{
+			name:  "ignore deferred",
```

Comment:
This one fails with the old logic, we missed it because the tests were only comparing 1 revisions struct for deferred scenarios which masked the bug

Reply:
🤦 🤦 🤦